frontend/src/metabase/query_builder/components/view/NativeQueryPreviewSidebar/NativeQueryPreviewSidebar.tsx

```diff
@@ -0,0 +1,124 @@
+import cx from "classnames";
+import { useCallback } from "react";
+import AceEditor from "react-ace";
+import { t } from "ttag";
+
+import { useGetNativeDatasetQuery } from "metabase/api";
+import { DelayedLoadingSpinner } from "metabase/common/components/EntityPicker/components/LoadingSpinner";
+import { color } from "metabase/lib/colors";
+import { formatNativeQuery, getEngineNativeType } from "metabase/lib/engine";
+import { useDispatch, useSelector } from "metabase/lib/redux";
+import { checkNotNull } from "metabase/lib/types";
+import { updateQuestion, setUIControls } from "metabase/query_builder/actions";
+import { NativeQueryEditorRoot } from "metabase/query_builder/components/NativeQueryEditor/NativeQueryEditor.styled";
+import { getQuestion } from "metabase/query_builder/selectors";
+import { Box, Button, Flex, Icon, rem } from "metabase/ui";
+import * as Lib from "metabase-lib";
+
+import SB from "./NativeQueryPreviewSidebar.module.css";
+import { createDatasetQuery } from "./utils";
+
+const TITLE = {
+  sql: t`SQL for this question`,
+  json: t`Native query for this question`,
+};
+
+const BUTTON_TITLE = {
+  sql: t`Convert this question to SQL`,
+  json: t`Convert this question to a native query`,
+};
+
+export const NativeQueryPreviewSidebar = (): JSX.Element => {
+  const dispatch = useDispatch();
+  const question = checkNotNull(useSelector(getQuestion));
+
+  const engine = question.database()?.engine;
+  const engineType = getEngineNativeType(engine);
+
+  const sourceQuery = question.query();
+  const canRun = Lib.canRun(sourceQuery);
+  const payload = Lib.toLegacyQuery(sourceQuery);
+  const { data, error, isLoading } = useGetNativeDatasetQuery(payload);
+
+  const showError = canRun && error;
```

Comment:
We probably should not show cached errors, i.e. error should not be shown when `error && isFetching`

https://github.com/metabase/metabase/assets/6830683/c83f530d-c2f6-4007-bd2a-5578eb7b9f5d